core/src/main/java/com/linecorp/armeria/internal/logging/ContentPreviewerConfigurator.java

```diff
@@ -0,0 +1,174 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.logging;
+
+import static java.util.Objects.requireNonNull;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Subscriber;
+
+import com.linecorp.armeria.common.FilteredHttpRequest;
+import com.linecorp.armeria.common.FilteredHttpResponse;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.logging.ContentPreviewer;
+import com.linecorp.armeria.common.logging.ContentPreviewerFactory;
+import com.linecorp.armeria.common.logging.RequestLogBuilder;
+import com.linecorp.armeria.internal.ArmeriaHttpUtil;
+
+/**
+ * A configurator to set up request and response content previewers.
+ */
+public final class ContentPreviewerConfigurator {
+
+    @Nullable
+    private final ContentPreviewerFactory requestContentPreviewerFactory;
+
+    @Nullable
+    private final ContentPreviewerFactory responseContentPreviewerFactory;
+
+    /**
+     * Creates a new instance.
+     */
+    public ContentPreviewerConfigurator(@Nullable ContentPreviewerFactory requestContentPreviewerFactory,
+                                        @Nullable ContentPreviewerFactory responseContentPreviewerFactory) {
+        assert requestContentPreviewerFactory != null || responseContentPreviewerFactory != null;
+        this.requestContentPreviewerFactory = requestContentPreviewerFactory;
+        this.responseContentPreviewerFactory = responseContentPreviewerFactory;
+    }
+
+    /**
+     * Sets up the request {@link ContentPreviewer} when {@code requestContentPreviewerFactory}
+     * is non-null and {@link ContentPreviewerFactory#get(RequestContext, HttpHeaders)} returns a
+     * {@link ContentPreviewer} instance that is not {@link ContentPreviewer#isDisabled()}.
+     */
+    public HttpRequest maybeSetUpRequestContentPreviewer(RequestContext ctx, HttpRequest req) {
+        requireNonNull(ctx, "ctx");
+        requireNonNull(req, "req");
+        if (requestContentPreviewerFactory == null) {
+            return req;
+        }
+
+        final RequestHeaders headers = req.headers();
+        final ContentPreviewer requestContentPreviewer = requestContentPreviewerFactory.get(ctx, headers);
+        if (requestContentPreviewer.isDisabled()) {
+            return req;
+        }
+
+        final RequestLogBuilder logBuilder = ctx.logBuilder();
+        logBuilder.deferRequestContentPreview();
+        requestContentPreviewer.onHeaders(headers);
+        req.completionFuture().handle((unused, unused1) -> {
```

Comment:
Please merge master and use `whenComplete()`.

Reply:
Fixed. 😉